src/Discovery/CollaboraDiscoveryFetcher.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+declare(strict_types=1);
+
+namespace Drupal\collabora_online\Discovery;
+
+use Drupal\collabora_online\Exception\CollaboraNotAvailableException;
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Config\ImmutableConfig;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\RequestOptions;
+use Psr\Http\Client\ClientExceptionInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\DependencyInjection\Attribute\Autowire;
+use Symfony\Component\ErrorHandler\ErrorHandler;
+
+/**
+ * Creates a discovery value object.
+ */
+class CollaboraDiscoveryFetcher implements CollaboraDiscoveryFetcherInterface {
+
+  public const DEFAULT_CID = 'collabora_online.discovery';
+
+  public function __construct(
+    #[Autowire(service: 'logger.channel.collabora_online')]
+    protected readonly LoggerInterface $logger,
+    protected readonly ConfigFactoryInterface $configFactory,
+    protected readonly ClientInterface $httpClient,
+    #[Autowire(service: 'cache.default')]
+    protected readonly CacheBackendInterface $cache,
+    #[Autowire(value: self::DEFAULT_CID)]
+    protected readonly string $cid,
+    protected readonly TimeInterface $time,
+  ) {}
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDiscovery(): CollaboraDiscoveryInterface {
+    $xml = $this->getDiscoveryXml();
+    $parsed_xml = $this->parseXml($xml);
+    return new CollaboraDiscovery($parsed_xml);
+  }
+
+  /**
+   * Parses an XML string.
+   *
+   * @param string $xml
+   *   XML string.
+   *
+   * @return \SimpleXMLElement
+   *   Parsed XML.
+   *
+   * @throws \Drupal\collabora_online\Exception\CollaboraNotAvailableException
+   *   The XML is invalid or empty.
+   */
+  protected function parseXml(string $xml): \SimpleXMLElement {
+    try {
+      // Avoid errors from XML parsing hitting the regular error handler.
+      // An alternative would be libxml_use_internal_errors(), but then we would
+      // have to deal with the results from libxml_get_errors().
+      $parsed_xml = ErrorHandler::call(
+        fn () => simplexml_load_string($xml),
+      );
```

Comment:
I just noticed `symfony/error-handler` package is required from "drupal/core-recommended" and from "symfony/http-kernel", but this means it is an indirect dependency and we should not rely on it. We should require it explicitly.

Reply:
nevermind we already do require it..